Simplify package mangement

mattijn opened the pull request:
Fix https://github.com/altair-viz/altair/issues/2939.

This PR intends to simplify the package management by moving to a complete pyproject.toml-based build. Meaning we can remove `setup.py`. I've adopted `hatch` as back-end. Reason: eg. jupyter-ecosystem use this. 
I also want to remove `setup.cfg`, meaning we have to upgrade from flake8 to another linter (flake8 has no pyproject.toml support). I've adopted `ruff` as new linter. Reason: eg. pandas and pylint use this.

I tried to use the same settings as our current flake8, but `ruff` finds still 57 errors. What to do with them? Fix in this PR?

Notes during exploration for this PR:

- Test-building the package can be done using `hatch build`. 
- `hatch shell` creates a pipenv based on the dependencies (no need for conda envs). 
- Optional dependencies can be installed using `pip install -e .[dev,doc]` (configured in pyproject.toml)
- I don't know yet what `hatch run` can do for us.

mattijn commented:
Any suggestions how to approach these ruff-raised errors @binste, @joelostblom?

joelostblom commented:
Cool, I haven't heard about `hatch` before, but I have had issues with other modern packaging tools like `poetry` which is why I have been sticking to setup.py myself. But `hatch` looks promising and it seems like it supports editable installs, which is necessary imo, but not all tools support for some reason.

Regarding the new syntax error raised by ruff, do they look like things you think we should fix? If so, I think they can be included here if they are relatively straightforward, but if there are larger changes maybe a separate PR with them would be easier to review (but it would still nice to see that PR, before merging this so that we have an idea of how complex the changes would be and if we run into something that is hard to fix).

mattijn commented:
Thanks @joelostblom, good suggestion. Will do!

binste commented:
Great PR! hatch seems like a good choice to me as well based on the adoption you mentioned, it being an official pypa project and hatchling is even the default build backend in the [Packaging Projects Python Guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/). I've only used it as a build backend, never the CLI, but it looks convenient too. We could migrate the 2 make files for running the tests and linters as well as building the docs to `hatch run` to further slim down the repository.

Also in favour of switching to ruff. The linting errors seem worth fixing/addressing. Let me know if I can help with that, migrating the make files, or reviewing this PR.

mattijn commented:
I'm not sure if this is too much, but I've deleted the files `setup.py`, `setup.cfg`, `requirements.txt`, `requirements_dev.txt`, `makefile` and `manifest.in`.

Instead of 
```cmd
make install && make test
```
one now can do
```
hatch run test
```
this will automatically use the development dependencies
furthermore the following are also supported
```cmd 
hatch run test-coverage
hatch run test-coverage-html
```
The makefile for building the docs is quite large. @binste, if you have any suggestions how to include this in the `pyproject.toml` file, that would be appreciated. Otherwise I think I'll skip it for now.

One thing to consider when working with `hatch.envs.<name>.scripts`, is that it is based on a new environment for scripts you want to execute. I now have used the `default` tag, to define an environment for the development dependencies:
```cmd
[tool.hatch.envs.default]
features = ["dev"]
```
in that case there is no need to name the environment in `hatch run test`, otherwise you will have to do `hatch run <name>:test`. For doc related scripts this will required.

binste commented:
I don't think it's too much at all, I really like this PR. Makes the repo much more organised, it's clearer where configurations and dependencies are specified (everything in `pyproject.toml`), one interface (`hatch` CLI) to interact with the project, everyone get's to benefit from running tests in isolated environments with automatically synced dependencies no matter how their local development setup looks, etc. 


From the docs makefile, only a few commands seem relevant to me for this project. They could be included as:

```toml

[tool.hatch.envs.doc]
features = ["dev", "doc"]

[tool.hatch.envs.doc.scripts]
clean = "rm -rf doc/_build"
clean-generated = [
    "rm -rf doc/user_guide/generated",
	"rm -rf doc/gallery"
]
clean-images = "rm -rf doc/_images"
clean-all = [
    "clean",
    "clean-generated",
    "clean-images"
]
html = [
    "mkdir -p doc/_images",
    "sphinx-build -b html -d doc/_build/doctrees doc doc/_build/html"
]
doctest = "sphinx-build -b doctest -d doc/_build/doctrees doc doc/_build/doctest"
coverage = "sphinx-build -b coverage -d doc/_build/doctrees doc doc/_build/coverage"
serve = "(cd doc/_build/html && python -m http.server)"
```



Tested all commands locally. The serve command is new. So to do a clean build of the docs and then serve them, one can do `hatch run doc:clean-all`, `hatch run doc:html`, `hatch run doc:serve`.

`doc/make.bat` and  `doc/Makefile` should then be removed. Note that the `clean` commands won't work on Windows as they use `rm`. Do you see a need to include Windows support as well? Building the docs might work but I can't test it.

binste commented on pyproject.toml:
```diff
@@ -1,6 +1,110 @@
+
+[tool.hatch.envs.default.scripts]
+test = [
+    "black --diff --color --check --line-length 88 .",
```
```suggestion
    "black --diff --color --check .",
```
Specifying the line length should not be necessary as it's the default for black

mattijn commented:
Thanks!  I included the scripts you provided for building the docs. I renamed the script name `html` to `build-html`, so it becomes `hatch run doc:build-html`.

I have removed the files `doc/make.bat`, `doc/makefile`, `doc/requiremtents.txt` as well and added a, sort of, table of content on top of the `pyproject.toml` to give a bit more clarity of the content of the file.

Last bit, is rewriting the `CONTRIBUTING.md` and `RELEASING.md` sections..

mattijn commented:
Thanks for the rolling review @binste! I've marked this PR now officially as _ready for review_, 

In the last changes I:
- Updated (simplified) the contributing.md and releasing.md files. 
- Generalized the `clean-all` script. This is different to what you had suggested beofre, is that OK?
- Included windows specific scripts for building the docs to provide options for a wider userbase (currently I'm on windows, so me included).

mattijn commented:
@joelostblom, if you have any comments or suggestions, please feel free to add them as well!

binste commented on pyproject.toml:
```diff
@@ -1,6 +1,153 @@
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[project]
```
Although when inspecting the source distribution it looks fine so maybe this could also be tested when uploading the `rc2` to pypi

binste reviewed (changes requested):
I only added some minor suggestions. Everything else looks good to me so feel free to merge afterwards!

I also tested the build process and installed the wheel file in another environment. A more thorough test is then probably the release of `rc2`.

mattijn commented on pyproject.toml:
```diff
@@ -1,6 +1,153 @@
+features = ["dev", "doc"]
+
+[tool.hatch.envs.doc.scripts]
+clean-all = [
```
I've included run scripts for `clean`  and `clean-generated` (and `-win` for windows-style).

mattijn commented:
Merging this🥳

joelostblom commented:
Great, thanks for all the work on making simplifying this! Unfortunately, I have limited time to review/contribute for the next ~3 weeks, so don't let me delay merging if someone else is available to review.

Narrat commented:
https://github.com/altair-viz/altair/blob/f3ced05897cb1e17f7299086436c0a08348c7028/pyproject.toml#L97-L99
Is it necessary to include `LICENSE`, `pyproject.toml` and `README.md` in that place? If a package for distribution is created, those three files will end up in `/usr/lib/python3.11/site-packages/` which doesn't seem to be the right place as they can be overwritten by other packages which could be place files there by error.
And in case of the license it would be a duplicate? Because it will be installed into `/usr/lib/python3.11/site-packages/altair-5.0.0.dist-info/licenses/`

binste commented:
Thank you @Narrat for bringing this up! You can follow the progress on resolving it at #3057.
